packages/devtools_app/lib/src/inspector/inspector_tree_controller.dart

```diff
@@ -542,6 +585,151 @@ class InspectorTreeController extends Object {
+  void previousMatch() {
+    super.previousMatch();
+    final currentSearchMatches = searchMatches.value;
+    for (int i = 0; i < currentSearchMatches.length; i++) {
+      final searchMatch = currentSearchMatches[i];
+
+      if (searchMatch.isSelected) {
+        final rowToSelect = i == 0
+            ? currentSearchMatches[currentSearchMatches.length - 1]
+            : currentSearchMatches[i - 1];
+
+        onSelectRow(rowToSelect);
+        break;
+      }
+    }
+  }
+
+  @override
+  void nextMatch() {
+    super.nextMatch();
+    final currentSearchMatches = searchMatches.value;
+    for (int i = 0; i < currentSearchMatches.length; i++) {
+      final searchMatch = currentSearchMatches[i];
+
+      if (searchMatch.isSelected) {
+        final rowToSelect = i == currentSearchMatches.length - 1
+            ? currentSearchMatches[0]
+            : currentSearchMatches[i + 1];
+
+        onSelectRow(rowToSelect);
+        break;
+      }
+    }
+  }
+
+  void dispose() {
+    if (_searchDebounce?.isActive ?? false) {
+      _searchDebounce.cancel();
+    }
+  }
+
+  Future _searchOperation;
+  Timer _searchDebounce;
+
+  @override
+  List<InspectorTreeRow> matchesForSearch(String search) {
+    updateMatches([]);
+    searchInProgress = true;
+
+    if (_searchDebounce?.isActive ?? false) {
+      _searchDebounce.cancel();
+    }
+    _searchDebounce =
+        Timer(Duration(milliseconds: search.isEmpty ? 0 : 300), () async {
+      // do something with query
+
+      // Abort any ongoing search operations and start a new one
+      if (_searchOperation != null) {
+        try {
+          await _searchOperation;
+        } catch (e) {
+          log(e, LogLevel.error);
+        }
+      }
+      searchInProgress = true;
+
+      // Start new search operation
+      _searchOperation = matchesForSearchAsyncOld(search, _searchTarget);
+
+      await _searchOperation;
+      searchInProgress = false;
+    });
+    return [];
+  }
+
+  Future<void> matchesForSearchAsyncOld(
+      String search, SearchTargetType searchTarget) async {
+    final matches = <InspectorTreeRow>[];
+
+    bool _firstSearchMatch = true;
+    void _updateSearchMatches() {
+      updateMatches(matches);
+
+      if (matches.isNotEmpty && _firstSearchMatch) {
+        _firstSearchMatch = false;
+        // Always select the first search match
+        onSelectRow(matches[0]);
+      }
+    }
+
+    int _statsSearchOps = 0;
+    int _statsWidgets = 0;
+
+    if (search == null ||
+        search.isEmpty ||
+        serviceManager.inspectorService == null ||
+        serviceManager.inspectorService.isDisposed) {
+      _updateSearchMatches();
+      debugPrint('Search completed, no search');
+      return;
+    }
+
+    debugPrint('Search started: ' + _searchTarget.toString());
+    final caseInsensitiveSearch = search.toLowerCase();
+
+    // Reset search matches
+    for (final row in cachedRows) {
+      _statsWidgets++;
+      setSearchMatch(row.node, false);
```

Comment:
we shouldn't need to do this. SearchControllerMixin already resets the isSearchMatch field in _updateMatches

Reply:
`setSearchMatch` does two things: it expands current node and it marks a `node` (not row) as search match. 
`InspectorTree` uses `getRow()` to get row data which is different from what's returned from `getCachedRow()`. `_updateMatches()` marks a row as search match but this row is a cached row so `isSearchMatch` is not reflecting the actual status for rows returned by `getRow()`.

I have the following questions regarding this:
* can cached rows be used in `InspectorTree`?
* what would be a good way to expand a node if `setSearchMatch()` is removed?

Reply:
A couple questsions:
1. Why do we need to mark both the row and the node as having a search match? as far as I can tell, everywhere `hasSearchMatch` is used, we have access to the row, so we could just write `row.isSearchMatch` instead.
2. Why are we expanding nodes at all for search?

If I understand correctly, the issue with cached rows vs non-cached rows is that we are setting `isSearchMatch` on cached rows and not on rows returned by `getRow()` - is that correct? If so, can we do a comparison where ever we need to call `getRow()` and need to know the value of `isSearchMatch` to check if there is a matching cached row with `isSearchMatch` set to true?

Reply:
>Why do we need to mark both the row and the node as having a search match?

It was made as a work around because `isSearchMatch` is set on cached rows but different instances of rows are used when displaying the tree.

>Why are we expanding nodes at all for search?

Search is looking for matches in `cachedRows`. In order to have all rows in `cachedRows` list all nodes need to be expanded.

>If I understand correctly, the issue with cached rows vs non-cached rows is that we are setting isSearchMatch on cached rows and not on rows returned by getRow() - is that correct?

Yes, that's correct.

>can we do a comparison where ever we need to call getRow() and need to know the value of isSearchMatch

Yes. We need to know the value of `isSearchMatch` in `_InspectorTreeState` but `getRows()` is used there. 

I've made some changes to the code:
* `setSearchMatch()` and `hasSearchMatch` have been removed
* `InspectorTree` now uses `getCachedRow()`. It should be okay to use it since it may still call `root?.getRow()` under the hood.

There's still need to expand the tree to get all rows cached. I did this by expanding all nodes of the tree in the overridden `set search` setter (so nodes are expanded before submitting search query). I don't think it's ideal, maybe it's better to find a solution that keeps current state of collapsed and expanded nodes.